### ISSUE TITLE
No longer resolving file paths correctly -  Error: ENOENT: no such file or directory, scandir

### ISSUE THREAD

**Issue by rlee89** [knex/knex#3660]
# Environment

Knex version: 0.20.9
Database + version: Postgres 10.6
OS: debian

# Bug

1. Since the latest version of Knex (0.20.9), once knex has switched to the working directory. Knex doesn't resolve the full path correctly. From the logs, I can see that Knex is trying to resolve the files from, for example, `/app/migrations` when in fact our folder structure is:
```
/app/config/knexfile.js
/app/config/migrations/*.js
/app/config/seed/*.js

```
2. 
```
event-service_1    | DB_MIGRATE set to 'true' (default is true)
event-service_1    | Running db migrations...
event-service_1    | 
event-service_1    | > event-service@1.0.0 db:migrate /app
event-service_1    | > knex --knexfile ./config/knexfile.js migrate:latest
event-service_1    | 
event-service_1    | Working directory changed to /app/config
event-service_1    | Error: ENOENT: no such file or directory, scandir '/app/migrations'
event-service_1    | npm ERR! code ELIFECYCLE
event-service_1    | npm ERR! errno 1
event-service_1    | npm ERR! event-service@1.0.0 db:migrate: `knex --knexfile ./config/knexfile.js migrate:latest`
event-service_1    | npm ERR! Exit status 1
event-service_1    | npm ERR! 
event-service_1    | npm ERR! Failed at the event-service@1.0.0 db:migrate script.
event-service_1    | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
event-service_1    | 
event-service_1    | npm ERR! A complete log of this run can be found in:
event-service_1    | npm ERR!     /home/node/.npm/_logs/2020-02-11T19_29_33_262Z-debug.log
event-service_1    | Completed db migrations
event-service_1    | DB_SEED set to 'true' (default is false)
event-service_1    | Running db seeding...
event-service_1    | 
event-service_1    | > event-service@1.0.0 db:seed /app
event-service_1    | > knex --knexfile ./config/knexfile.js seed:run
event-service_1    | 
event-service_1    | Working directory changed to /app/config
event-service_1    | Error: ENOENT: no such file or directory, scandir '/app/seeds'
```
3. 
Looks to be related to changes here: https://github.com/knex/knex/commit/947273e24a73545a269fbe90ef30d55889d5db72

Thanks for the help! 
Richard

**Comment by kibertoad** [comment]
@briandamaged

**Comment by briandamaged** [comment]
@rlee89 : Howdy, and sry about that!

I'm attempting to recreate the issue, but I have not had any success so far.  I setup a standalone project w/ the following structure:

```
.
└── foo
    ├── config
    │   ├── knexfile.js
    │   └── migrations
    │       ├── 20200212111031_foo.js
    │       └── 20200212113040_blarg.js
    ├── package-lock.json
    └── package.json
```

When I run `knex` from within the `foo/` folder, I receive the following output:

```
 % ./node_modules/.bin/knex --knexfile=./config/knexfile.js  migrate:latest
Working directory changed to ~/tmp/foo/config
Using environment: development
sqlite does not support inserting default values. Set the `useNullAsDefault` flag to hide this warning. (see docs http://knexjs.org/#Builder-insert).
Batch 1 run: 2 migrations
```

This seems to match the behavior that you were expecting to see... 🤔 

So... here are a few other thoughts:

1. Perhaps it's a relative path issue within the `knexfile.js` itself?  Ex: is the `knexfile.js` specifying any paths for the migration or seed folders?
2. What directory is the shell in when invoking the `knex` command? (In theory, this shouldn't make any difference, but you never know)
3. Is the `KNEX_PATH` environment variable being set anywhere?
4. Is the `--cwd` flag being passed into the `knex` command?  (The output provided suggests that this is not the case)

I'll continue digging through the code to see if I can identify any other quirks that might be causing the issue.

**Comment by briandamaged** [comment]
@rlee89 : Hi again!  Can you please include the NPM debug log as well?  For the run you posted earlier, this log was located at:

```
/home/node/.npm/_logs/2020-02-11T19_29_33_262Z-debug.log
```

Based upon what I'm seeing in the code so far, my guess is that one of the following is happening:

1. Your `knexfile.js` is creating a custom MigrationSource.  This MigrationSource is attempting to resolve some paths before the process's current working directory has been changed.
2. Another plugin/hook/whatever is resetting the process's current working directory back to `/app`.  This is then causing the relative path (`./migrations`) to resolve to the wrong folder.

@kibertoad : The more I think about it, the more I think that `knex` probably shouldn't be modifying the process's current working directory at all.  Instead, it probably should provide custom logic for resolving relative paths.  This would allow `knex` to have consistent behavior regardless of how the developer's code manipulates the process's current working directory.  Thoughts?

**Comment by briandamaged** [comment]
@kibertoad : To clarify my earlier point: from the end-user's perspective, Knex would still retain its current behavior.  But, its implementation would no longer be tightly coupled to `process.cwd()`.  Thus, even if the value of `process.cwd()` gets changed after Knex was initialized, Knex would still be able to resolve paths correctly.

**Comment by kibertoad** [comment]
Yeah, I agree, that would indeed be an improvement.

**Comment by rlee89** [comment]
Thanks @briandamaged for the quick reply and investigation! 

> 1. Perhaps it's a relative path issue within the knexfile.js itself? Ex: is the knexfile.js specifying any paths for the migration or seed folders?

here are the first couple lines of initializing any paths:
```
'use strict';
const path = require('path');
const MIGRATION_DIR = path.resolve('migrations');
const SEED_DIR = path.resolve('seeds');
const MIGRATION_TABLE_NAME = 'migrations';
```

> 2. What directory is the shell in when invoking the knex command? (In theory, this shouldn't make any difference, but you never know)

We're running in the /app directory. Your standalone projects structure is similar to how we've laid out our folder structure. 

> 3. Is the `KNEX_PATH` environment variable being set anywhere?

Nope

> 4. Is the --cwd flag being passed into the knex command? (The output provided suggests that this is not the case)

You're correct, not the case.

Attaching the requested logs (different date because the output dir for those logs were in a container and long gone).
[2020-02-12T21_44_28_968Z-debug.log](https://github.com/knex/knex/files/4195231/2020-02-12T21_44_28_968Z-debug.log)

> Based upon what I'm seeing in the code so far, my guess is that one of the following is happening:
> 
> Your knexfile.js is creating a custom MigrationSource. This MigrationSource is attempting to resolve some paths before the process's current working directory has been changed.
> Another plugin/hook/whatever is resetting the process's current working directory back to /app. This is then causing the relative path (./migrations) to resolve to the wrong folder.

It does seem odd, though pinning the knex version back to `0.20.8` with no other changes seems to work fine. I suspect it does have something to do with where the shell invokes the command. 

```
event-service_1    | wait-for-it.sh: postgres:5432 is available after 4 seconds
event-service_1    | DB_MIGRATE set to 'true' (default is true)
event-service_1    | Running db migrations...
event-service_1    | 
event-service_1    | > event-rule-service@1.0.0 db:migrate /app
event-service_1    | > knex --knexfile ./config/knexfile.js migrate:latest
event-service_1    | 
event-service_1    | Batch 1 run: 5 migrations
event-service_1    | Completed db migrations
event-service_1    | DB_SEED set to 'true' (default is false)
event-service_1    | Running db seeding...
event-service_1    | 
event-service_1    | > event-rule-service@1.0.0 db:seed /app
event-service_1    | > knex --knexfile ./config/knexfile.js seed:run
```
Appreciate the help guys!

**Comment by briandamaged** [comment]
@rlee89 : Okay, I think I see the issue now.  The sequence of operations within the Knex CLI is:

1. Open the `knexfile.js`
2. Change the process's current working directory
3. Attempt to run the migrations / etc.

Your `knexfile.js` is resolving some paths as soon as it's opened.  The problem: they are being resolved _before_ Knex has changed the process's current working directory.  This is causing the paths to be calculated relative to `/app` instead of `/app/config`.

So, I'll poke around a bit and see if I can fix the order of operations on the Knex side.  As for a short-term solution: try removing the calls to `path.resolve(..)` from your `knexfile.js`.  This will cause the paths to remain relative until the moment that Knex attempts to run the migrations / etc .  By then, the process's current working directory will already be set to the correct location.

**Comment by briandamaged** [comment]
@rlee89 : Actually, this sorta relates to the earlier conversation between @kibertoad and I.

Generally speaking, it's risky for modules to make assumptions about the current value of `process.cwd()`.  It can lead to all sorts of surprises when another piece of code changes the process's current working directory.

If possible, you should try to resolve paths relative to a fixed value, such as `__dirname`.  For example:

```
// __dirname   will always be the folder that contains  knexfile.js
const MIGRATION_DIR = path.resolve(__dirname, 'migrations');
```

**Comment by rlee89** [comment]
> @rlee89 : Okay, I think I see the issue now. The sequence of operations within the Knex CLI is:
> 
> 1. Open the `knexfile.js`
> 2. Change the process's current working directory
> 3. Attempt to run the migrations / etc.
> 
> Your `knexfile.js` is resolving some paths as soon as it's opened. The problem: they are being resolved _before_ Knex has changed the process's current working directory. This is causing the paths to be calculated relative to `/app` instead of `/app/config`.
> 
> So, I'll poke around a bit and see if I can fix the order of operations on the Knex side. As for a short-term solution: try removing the calls to `path.resolve(..)` from your `knexfile.js`. This will cause the paths to remain relative until the moment that Knex attempts to run the migrations / etc . By then, the process's current working directory will already be set to the correct location.

Yup, we look good! Removed the `path.resolve(...)` and that worked. We'll look at seeing if can't suggest cleaning that up for all our services too... 😄 

```
event-service_1    | DB_MIGRATE set to 'true' (default is true)
event-service_1    | wait-for-it.sh: postgres:5432 is available after 4 seconds
event-service_1    | Running db migrations...
event-service_1    | 
event-service_1    | > event-service@1.0.0 db:migrate /app
event-service_1    | > knex --knexfile ./config/knexfile.js migrate:latest
event-service_1    | 
event-service_1    | Batch 1 run: 5 migrations
event-service_1    | Completed db migrations
event-service_1    | DB_SEED set to 'true' (default is false)
event-service_1    | Running db seeding...
event-service_1    | 
event-service_1    | > event-service@1.0.0 db:seed /app
event-service_1    | > knex --knexfile ./config/knexfile.js seed:run
event-service_1    | 
event-service_1    | Ran 1 seed files
event-service_1    | Completed db seeding
event-service_1    | 
event-service_1    | > event-service@1.0.0 start /app
event-service_1    | > node app.js
event-service_1    | 
```

**Comment by kibertoad** [comment]
@rlee89 Fix released on 0.20.10.
